### PR TITLE
8228343: JCMD and attach fail to work across Linux Container boundary

### PR DESCRIPTION
**Pull request by kevinjwalls**
Since 8214300, jcmd cannot attach to a Java process in a docker container.

That change started using a canonicalized File to create the .attach_pidXXX file.  For a target process in a container, it will follow a symlink that is likely not the same as for the target process.  e.g. follow a symlink to a cwd of / which is not the same directory for the container host, as it is within the container.  Containerized VM never sees the file, never creates the socket file, the attach times out and fails.

To keep the 8214300 change working for non-container situations, we can keep a canonical version of the attach File to use for deleting.

For containers there will remain the problem 8214300 describes, although it is unlikely: if you start the attach to a containerized VM,  and it then exits, we can't delete the .attach_pidXXX file.  Neither the /proc/PID/cwd or canonical form are any use.

(Possibly leaving a .attach_pidXXX file if the target dies in that small window is better than the current situation.)

Here I'm suggesting the same change on AIX, although I can't build/test that.  I'm expecting it has the same problem, as /proc/pid/cwd is still a symlink.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8228343](https://bugs.openjdk.java.net/browse/JDK-8228343): JCMD and attach fail to work across Linux Container boundary

### Reviewers
 * [Yasumasa Suenaga](https://openjdk.java.net/census#ysuenaga) (@YaSuenag - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4418/head:pull/4418` \
`$ git checkout pull/4418`

Update a local copy of the PR: \
`$ git checkout pull/4418` \
`$ git pull https://git.openjdk.java.net/jdk pull/4418/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4418`

View PR using the GUI difftool: \
`$ git pr show -t 4418`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4418.diff">https://git.openjdk.java.net/jdk/pull/4418.diff</a>

</details>

### CODE REVIEW

**Comment by mseledts** [comment]
I have verified the fix on Linux-x64 (OL 7.9). It works now. Thank you for fixing this.

**Comment by kevinjwalls** [comment]
Thanks Mikhailo for verifying.

**Comment by YaSuenag** [comment]
Can we use `/proc/<PID>/root`? According to [manpage](https://man7.org/linux/man-pages/man5/proc.5.html), I think we can access `/` in the container for sure.
This directory is used in `findSocketFile()`.
If it can't be used, you might have to change it too.

**Comment by kevinjwalls** [comment]
Thanks Yasumasa --

> Can we use /proc/<PID>/root? 

root is also a symlink, like cwd, so we can't resolve its canonical location (literally "/") and put a file there, or if our target is in a container it won't see it.  So the problem is not that we can't access /proc/PID/cwd/ but it's that if using the canonical filename, jcmd and its target will resolve them differently.

Maybe it is interesting that we don't use cwd for the socket file, but it's not something I would want to start changing at this time... 8-)

**Comment by kevinjwalls** [comment]
I do see the manpage section about root, saying "this file is not merely a symbolic link", but it does behave like a regular symlink for me...   Using File getCanonicalFile() on it returns me a literal "/" root directory File.

File: src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java [hunk 1]
```diff
@@ -77,6 +77,8 @@
         socket_path = socket_file.getPath();
         if (!socket_file.exists()) {
             File f = createAttachFile(pid, ns_pid);
+            // Keep a canonical version of File, to delete, in case target process ends and /proc link has gone:
+            File f2 = f.getCanonicalFile();
```
Comment (YaSuenag) [inline src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java:81]: Do we need to keep `f`? I think we can replace them to `createAttachFile(pid, ns_pid).getCanonicalFile()`.

**Review by YaSuenag — COMMENTED** [review]
Thanks @kevinjwalls ! I thought we can get real location from `/proc/<PID>/root`, however it do not so. I confirmed it on podman. (podman uses overlayfs, and I couldn't find an easy way to access it directly.

Your change almost looks good, but I have only one comment.

File: src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java (hunk 1 above)
Comment (kevinjwalls) [reply src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java]: That's true, we don't really need them both.

Comment (kevinjwalls) [reply src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java]: Builds and tests OK, thanks.

**Comment by YaSuenag** [comment]
I hope this change happens in 17.

**Review by sspitsyn — APPROVED** [review]
Hi Kevin,
This looks okay to me.
How do you test it to make sure no regressions are introduced?
Thanks,
Serguei

**Comment by kevinjwalls** [comment]
> Hi Kevin,
> This looks okay to me.
> How do you test it to make sure no regressions are introduced?
> Thanks,
> Serguei

Yes, the bug text here mentions JDK-8195809 which is an effort to create a docker test...  With this fixed, that can go ahead.  The updated test is attached there - I'm not yet sure exactly where it will live, or when it will run, but we do need it...

**Comment by kevinjwalls** [comment]
> I hope this change happens in 17.

Yes, I think we can make it!...  Thanks.

**Pull request closed without merging**